File: Task2/Code/procedural.cpp
//------------------------------------------------------------------------------
// procedural.cpp - содержит функции обработки процедурных языков
//------------------------------------------------------------------------------

#include "procedural.h"

// от 0 до 1 для булева значения
Random Procedural::rnd_bool(0, 1);

//------------------------------------------------------------------------------
// Ввод параметров процедурного языка из потока
void Procedural::In(ifstream &ifst) {
    string name;
    ifst >> name >> abstract_types >> popularity >> year;
    name_len = name.size();
}

// Случайный ввод параметров процедурного языка
void Procedural::InRnd() {
    name_len = rnd_len.GetInt();
    abstract_types = static_cast<bool>(rnd_bool.GetInt());
    popularity = rnd_popularity.GetDouble();
    year = rnd_year.GetInt();
}

//------------------------------------------------------------------------------
// Вывод параметров процедурного языка в поток
void Procedural::Out(ofstream &ofst) {
    ofst << "It is Procedural programming language: number of characters = "
         << name_len << ",  abstract types = "
         << boolalpha << abstract_types << noboolalpha;

    if (abstract_types) {
        ofst << ".\t\t\tPopularity = " << popularity;
    } else {
        ofst << ".\t\t\tPopularity = " << popularity;
    }

    ofst << "%, year = " << year << ", quotient = " << Quotient() << "\n";
}

//------------------------------------------------------------------------------
// Вычисление частного процедурного языка
double Procedural::Quotient() {
    return year / double(name_len);
}

File: Task3/Code/procedural.py
import random

from language import Language

# Класс Процедурного языка программирования
class Procedural(Language):
    def __init__(self):
        self.name_len = 0 # длина названия языка
        self.abstract_types = 0 # наличие абстрактных типов
        self.popularity = 0 # популярность
        self.year = 0 # год создания

    # Ввод данных в объект класса из строки файла
    def input(self, str):
        self.name_len = int(len(str[1]))
        self.abstract_types = bool(str[2])
        self.popularity = float(str[3])
        self.year = int(str[4])

    # Ввод данных в объект класса случайным образом
    def inputrnd(self):
        self.name_len = random.randint(1, 20)
        self.abstract_types = bool(random.randint(1, 2))
        self.popularity = round(random.random() * (100 - 1) + 1, 3)
        self.year = random.randint(1957, 2020)
        pass

    # Вывод данных объекта класса в файл
    def output(self, ostr):
        if self.abstract_types:
            self.abstract_types = "True.\t"
        else:
            self.abstract_types = "False."
        ostr.write(
            f"It is Procedural programming language: number of characters = {self.name_len},  "
            f"abstract types = {self.abstract_types}\t\t\tPopularity = {self.popularity}%, "
            f"year = {self.year}, quotient = {self.quotient()}\n")
        pass

    # Подсчет частного для данного объекта класса языка программирования
    def quotient(self):
        return round(self.year / float(self.name_len), 3)
